### PR TITLE
[poolkeeper] Introduce poolkeeper: a to hedge our dev clusters

### PR DESCRIPTION
**Pull request by geropl**
Poolkeeper is meant to take over the tedious tasks that come up with our dev clusters. It:
 - ensures that we properly scale down when no preview envs or builds are running
 - ensures that we have a running build VM when we need it

This enables us to use safely use bigger build nodes because they reliably scale down: https://github.com/gitpod-io/gitpod/pull/2541

Possible future tasks:
 - clean out things that pile up with preview envs due to the high throughput of Gitpod installations (containers, images, IDEs, etc.)
 - check for preview envs piling up (might be better located in sweeper?)

### CODE REVIEW

**Review by wulfthimm — APPROVED** [review]
LGTM but I cannot verify the Go code. As it is in use already, I give my consent.

File: dev/poolkeeper/leeway.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+# Licensed under the GNU Affero General Public License (AGPL).
+# See License-AGPL.txt in the project root for license information.
+
+FROM alpine:latest
+
+RUN apk add --no-cache ca-certificates curl git
+RUN curl -L https://github.com/csweichel/werft/releases/download/v0.0.4rc/werft-client-linux-amd64.tar.gz | tar xz && \
```
Comment (geropl) [inline dev/poolkeeper/leeway.Dockerfile:8]: Leftover

File: dev/poolkeeper/pkg/poolkeeper/poolkeeper.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+		stop: make(chan struct{}, 1),
+		done: make(chan struct{}, 1),
+	}
+}
+
+// Start starts the PoolKeeper and is meant to be run in a goroutine
+func (pk *PoolKeeper) Start() {
+	defer func() {
+		pk.done <- struct{}{}
```
Comment (csweichel) [inline dev/poolkeeper/pkg/poolkeeper/poolkeeper.go:40]:
```suggestion
		close(pk.done)
```

File: dev/poolkeeper/pkg/poolkeeper/task-keep-nodes-alive.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	t = t.Local()
+	tod := time.Date(0, 1, 1, t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), t.Location())
+	start := time.Time(k.PeriodStart)
+	end := time.Time(k.PeriodEnd)
+	if tod.Before(start) {
+		log.Debug("nothing to do")
+		return
+	} else if tod.After(start) && tod.Before(end) {
```
Comment (csweichel) [inline dev/poolkeeper/pkg/poolkeeper/task-keep-nodes-alive.go:62]:
```suggestion
	}
	if tod.After(start) && tod.Before(end) {
```

File: dev/poolkeeper/pkg/poolkeeper/task-keep-nodes-alive.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+					},
+				},
+				NodeSelector: k.NodeSelector,
+			},
+		})
+		if err != nil {
+			log.WithError(err).Errorf("error creating keep-alive pod")
+			return
+		}
+		log.WithField("pod", pod.Name).Info("created pod")
```
Comment (csweichel) [inline dev/poolkeeper/pkg/poolkeeper/task-keep-nodes-alive.go:95]:
```suggestion
		log.WithField("pod", pod.Name).Info("created pod")
		return
	}
	if tod.After(end) {
```

**Review by csweichel — COMMENTED** [review]
Code LGTM (a few very minor code nits).

I wonder though if we couldn't achieve the same tasks using [Kubernetes cron jobs](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/)? We might still have a CLI that implements the tasks (if `kubectl patch` does not do the trick already), but we would not have to roll our own cron service.

**Comment by geropl** [comment]
> I wonder though if we couldn't achieve the same tasks using Kubernetes cron jobs?

Fair point. I looked into that at first, but then wanted sth. that I feel better with than `yaml` + bash scripts: Go gives us perfect control without any major downside IMO. 

> We might still have a CLI that implements the tasks (if kubectl patch does not do the trick already), but we would not have to roll our own cron service

IMO that solution would be a net negative because we had the additional CornJob `yaml` complexity just for the timers (which are as simple as it can get in Go).

**Pull request merged**